superduperdb/ext/openai/model.py

```diff
@@ -71,18 +71,38 @@ def __post_init__(self):
         self.encoder = vector(self.shape)
 
     @retry
-    def _predict_one(self, X, **kwargs):
+    def predict_one(self, X: str, **kwargs):
         e = Embedding.create(input=X, model=self.identifier, **kwargs)
         return e['data'][0]['embedding']
 
+    @retry
+    async def apredict_one(self, X: str, **kwargs):
+        e = await Embedding.acreate(input=X, model=self.identifier, **kwargs)
+        return e['data'][0]['embedding']
+
     @retry
     def _predict_a_batch(self, texts, **kwargs):
         out = Embedding.create(input=texts, model=self.identifier, **kwargs)['data']
         return [r['embedding'] for r in out]
 
-    def _predict(self, X, batch_size=100, **kwargs):  # asyncio?
+    @retry
+    async def _apredict_a_batch(self, texts, **kwargs):
+        out = await Embedding.create(input=texts, model=self.identifier, **kwargs)[
+            'data'
+        ]
+        return [r['embedding'] for r in out]
+
+    def _predict(self, X, batch_size=100, **kwargs):
         if isinstance(X, str):
-            return self._predict_one(X)
+            return self.predict_one(X)
+        out = []
+        for i in tqdm.tqdm(range(0, len(X), batch_size)):
+            out.extend(self._predict_a_batch(X[i : i + batch_size], **kwargs))
+        return out
+
+    async def _apredict(self, X, batch_size=100, **kwargs):
+        if isinstance(X, str):
+            return await self.apredict_one(X)
         out = []
         for i in tqdm.tqdm(range(0, len(X), batch_size)):
             out.extend(self._predict_a_batch(X[i : i + batch_size], **kwargs))
```

Comment:
A comment for the future:

At the moment, we fire off the async calls in a synchronous loop ie we fire the call, and then the async event loop periodically checks to see if it has returned, and when it has we move on to fire the next call. 

We could also consider an async loop ie we fire off all the calls at once, and then periodically check in on each of them. There is a limit to this (the OpenAI rate limit, as you have already referenced in the code), but perhaps there is some scope for optimisation in the future here. (Or perhaps this is already addressed by the 'batches' idea.)

Reply:
To be revisited! Let's see if this meets our performance expectations, and potentially build this in.